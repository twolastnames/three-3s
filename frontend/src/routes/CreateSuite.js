import React from 'react';
import { useState } from 'react';
import { fetchWithMessages } from '../helpers/fetchWithMessages';
import PropTypes from 'prop-types';

const postWithMessages = fetchWithMessages({ method: 'POST' }, 'info');

const getInput = () => document.getElementById('create_suite_name');

const getName = () => (getInput() || {}).value;

const clearName = () => ((getInput() || {}).value = '');

const getUrl = () => `/threAS3/suites?name=${getName()}`;

const getDescription = () => `suite named '${getName()}'`;

const createSuite = async (post) => await post(getDescription(), getUrl());

const onClick = (post, setButtonDisabled) => (event) => {
  event.preventDefault();
  createSuite(post);
  clearName();
  setButtonDisabled(true);
};

const onChange = (setButtonDisabled) => ({ target: { value } }) =>
  setButtonDisabled(!value);

Input.propTypes = {
  setButtonDisabled: PropTypes.func.isRequired,
};

function Input({ setButtonDisabled }) {
  return (
    <input
      onChange={onChange(setButtonDisabled)}
      type="text"
      id="create_suite_name"
    />
  );
}

Button.propTypes = {
  buttonDisabled: PropTypes.bool.isRequired,
  post: PropTypes.func.isRequired,
  setButtonDisabled: PropTypes.func.isRequired,
};

function Button({ buttonDisabled, post, setButtonDisabled }) {
  return (
    <button
      disabled={buttonDisabled}
      onClick={onClick(post, setButtonDisabled)}
      value="Create"
    >
      Create
    </button>
  );
}

Form.propTypes = {
  buttonDisabled: PropTypes.bool.isRequired,
  post: PropTypes.func.isRequired,
  setButtonDisabled: PropTypes.func.isRequired,
};

function Form({ buttonDisabled, post, setButtonDisabled }) {
  return (
    <form>
      <label>
        Name
        <Input setButtonDisabled={setButtonDisabled} />
      </label>
      <Button
        buttonDisabled={buttonDisabled}
        post={post}
        setButtonDisabled={setButtonDisabled}
      />
    </form>
  );
}

export function CreateSuite() {
  const [buttonDisabled, setButtonDisabled] = useState(true);

  return (
    <div className='route'>
      <h2>Create Suite</h2>
      <Form
        buttonDisabled={buttonDisabled}
        post={postWithMessages}
        setButtonDisabled={setButtonDisabled}
      />
    </div>
  );
}
