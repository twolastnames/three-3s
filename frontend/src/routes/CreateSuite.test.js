import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';
import { CreateSuite } from './CreateSuite';

describe('CreateSuite', () => {
  it('renders visually', () => {
    expect(
      shallow(<CreateSuite displayMessage={() => {}} />)
    ).toMatchSnapshot();
  });

  describe('a successful response', () => {
    let displayMessage = jest.fn();

    beforeEach(async () => {
      jest.spyOn(global, 'fetch').mockImplementation(() =>
        Promise.resolve({
          ok: true,
          json: async () => {
            'data';
          },
        })
      );
      const wrapper = mount(<CreateSuite displayMessage={displayMessage} />);
      act(() =>
        wrapper.find('input').prop('onChange')({ target: { value: 'hello' } })
      );
      await nextTick();
      wrapper.find('button').simulate('click', { preventDefault: () => {} });
      await nextTick();
    });

    afterEach(() => {
      global.fetch.mockReset();
    });

    it('gives an info acknowledgement', () => {
      expect(displayMessage).toBeCalledTimes(1);
      expect(displayMessage).toBeCalledWith(
        'info',
        "success with suite named 'undefined'"
      );
    });

    it('is called with requested uri', () => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith('/threAS3/suites?name=undefined', {
        method: 'POST',
      });
    });
  });

  describe('an unsuccessful response', () => {
    let displayMessage = jest.fn();

    beforeEach(async () => {
      jest
        .spyOn(global, 'fetch')
        .mockImplementation(() => Promise.resolve({ status: 666 }));
      const wrapper = mount(<CreateSuite displayMessage={displayMessage} />);
      act(() =>
        wrapper.find('input').prop('onChange')({ target: { value: 'hello' } })
      );
      await nextTick();
      wrapper.find('button').simulate('click', { preventDefault: () => {} });
      await nextTick();
    });

    afterEach(() => {
      global.fetch.mockReset();
    });

    it('gives an error acknowledgement', () => {
      expect(displayMessage).toBeCalledTimes(1);
      expect(displayMessage).toBeCalledWith(
        'error',
        "suite named 'undefined' with HTTP code 666"
      );
    });

    it('is called with requested uri', () => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith('/threAS3/suites?name=undefined', {
        method: 'POST',
      });
    });
  });

  describe('an exception', () => {
    let displayMessage = jest.fn();

    beforeEach(async () => {
      jest.spyOn(global, 'fetch').mockImplementation(() => {
        throw new Error('a message');
      });
      const wrapper = mount(<CreateSuite displayMessage={displayMessage} />);
      act(() =>
        wrapper.find('input').prop('onChange')({ target: { value: 'hello' } })
      );
      await nextTick();
      wrapper.find('button').simulate('click', { preventDefault: () => {} });
      await nextTick();
    });

    afterEach(() => {
      global.fetch.mockReset();
    });

    it('gives an error acknowledgement', () => {
      expect(displayMessage).toBeCalledTimes(1);
      expect(displayMessage).toBeCalledWith(
        'error',
        "suite named 'undefined' a message"
      );
    });

    it('is called with requested uri', () => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith('/threAS3/suites?name=undefined', {
        method: 'POST',
      });
    });
  });
});
