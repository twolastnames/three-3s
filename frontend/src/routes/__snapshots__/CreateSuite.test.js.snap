// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSuite renders visually 1`] = `ShallowWrapper {}`;
